### PR TITLE
removes an explicit uplink component check in favor of a signal in detomatix carts

### PR DESCRIPTION
**Pull request by vuonojenmustaturska**
## About The Pull Request

This PR adds a new signal to do the "is this PDA detonatable" checking, then registers for that signal in the uplink component and also a new pda protection component used in captain and AI pda's. This has no gameplay changes, they were previously undetonateable as well.

## Changelog
:cl: Naksu
code: pda.detonatable and uplink component checks removed in favor of a signal in detomatix cart handling
/:cl:

### CODE REVIEW

**Comment by Fikou** [comment]
as far as i know you can blow the caps pda, did it like one month ago as nukie

**Comment by ninjanomnom** [comment]
Every GetComponent removed gives me more strength, very nice

**Comment by ATH1909** [comment]
> as far as i know you can blow the caps pda, did it like one month ago as nukie

The captain's PDA is marked as being immune in the code; if it's still detonate-able, that's a bug.

File: code/datums/components/pda_protection.dm [hunk 1]
```diff
@@ -0,0 +1,10 @@
+/datum/component/pda_protection
+	dupe_mode = COMPONENT_DUPE_UNIQUE_PASSARGS
+
+/datum/component/pda_protection/Initialize()
+	if(!istype(parent, /obj/item/pda))
+		return COMPONENT_INCOMPATIBLE
+	RegisterSignal(parent, list(COMSIG_PDA_CHECK_DETONATE), .proc/check)
+
+/datum/component/pda_protection/proc/check()
+	return COMPONENT_PDA_NO_DETONATE
```
Comment (ninjanomnom) [inline code/datums/components/pda_protection.dm]: This is better just as an object listening for the signal on itself

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/game/objects/items/devices/PDA/PDA_types.dm [hunk 2]
```diff
@@ -33,7 +32,9 @@
 /obj/item/pda/ai/pai
 	ttone = "assist"
 
-
+/obj/item/pda/ai/pai/Initialize()
```
Comment (YPOQ) [inline code/game/objects/items/devices/PDA/PDA_types.dm]: I think you wanted just `/obj/item/pda/ai/Initialize()`

File: code/game/objects/items/devices/PDA/PDA.dm [hunk 3]
```diff
@@ -1009,6 +1009,9 @@ GLOBAL_LIST_EMPTY(PDAs)
 			continue
 		. += P
 
+/obj/item/pda/proc/pda_no_detonate()
```
Comment (vuonojenmustaturska) [inline code/game/objects/items/devices/PDA/PDA.dm]: static callbacks when

File: code/game/objects/items/devices/PDA/PDA_types.dm (hunk 2 above)
Comment (vuonojenmustaturska) [reply code/game/objects/items/devices/PDA/PDA_types.dm]: i did

File: code/game/objects/items/devices/PDA/virus_cart.dm [hunk 4]
```diff
@@ -71,8 +71,7 @@
 				difficulty++ //if cartridge has manifest access it has extra snowflake difficulty
 			else
 				difficulty += 2
-		GET_COMPONENT_FROM(hidden_uplink, /datum/component/uplink, target)
-		if(!target.detonatable || prob(difficulty * 15) || (hidden_uplink))
+		if(SEND_SIGNAL(target, COMSIG_PDA_CHECK_DETONATE) == COMPONENT_PDA_NO_DETONATE || prob(difficulty * 15))
```
Comment (ninjanomnom) [inline code/game/objects/items/devices/PDA/virus_cart.dm]: It's a flag
```suggestion
		if(SEND_SIGNAL(target, COMSIG_PDA_CHECK_DETONATE) & COMPONENT_PDA_NO_DETONATE || prob(difficulty * 15))
```

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

**Review by ninjanomnom — APPROVED** [review]

**Pull request merged**